### ISSUE TITLE
Action button forced save of record breaks Edit - Save/Discard paradigm

### ISSUE THREAD

**Issue by elgow** [odoo/odoo#20942]
Impacted versions: 11.0 (and really 10 and 9, but let's not worry about those)

Steps to reproduce: 

1. Install a clean Odoo 11 and add a plugin that uses Mail, e.g. the Project plugin. 
1. Go to Settings/Email/Emails in the admin menu
1. Select one of the demo data emails colored red, e.g. "Reception of Create new components"
1. Hit _Edit_ to go to edit mode
1. Edit the message title
1. Press the _Retry_ button
1. Press _Cancel_ for the retry
1. Press _Discard_ for the edit

Note that the change you made to the message title has not been discarded.

Current behavior: The code in the function `web.FormController._onButtonClicked` absolutely forces an action button press to save the record w/o the user having pressed _Save_. In the case where an action button is used to contribute something to an edit, e.g. prepared text to insert into an email, the forced save does not respect the user's view of the edit session.  It is possible to warn the user of the save, but it's not really possible to prevent it. 

Expected behavior: The _Edit_ - _Save/Discard_ buttons lead the user to expect that the edit is not committed until _Save_ is pressed and that _Discard_ means that the edit will not be saved. The current behavior violates the user's reasonable expectations by saving w/o the user requesting it. It also prevents some very valuable uses of wizards invoked by action buttons within an edit session. 

Video/Screenshot link (optional):

**Comment by Yenthe666** [comment]
@ged-odoo looks like an interesting case for you to check.

**Comment by ged-odoo** [comment]
Dear All,

I understand the concern, and I actually kind of agree, but this is actually really completely a fundamental mechanism of how odoo works.  The reason is that many actions called by clicking on a button need to have many side effects, such as reading data in other models, sending emails, writing on various other records.  By saving first, we ensure that the state of the data in the database matches what the user is expecting. 

The alternative would be to send to the server the current state of the form view (like onchanges do), then performing the necessary side effects, then rollbacking only the changes coming from the form view. This would be probably quite complex (if even possible), and in many cases, not what the user would expect.

Maybe we could add an attribute to the button to explicitely declare if saving is necessary.  But this would make user and developer's life more complicated, for various reasons.

In short, this is really really difficult to do this in another way...  So, unless we have a really concrete proposition, I'll close this. Sorry, and thanks for taking the time to read

**Comment by elgow** [comment]
Actually you could add an attribute to the button, but make it declare that saving is NOT necessary. If not applied then it breaks nothing and changes nothing. If applied then it's on the developer to use it wisely. 

Doing the above is not difficult. I've already done it as an addon module and I am using it. It can be done even more simply if it's not a monkey-patch. Then it's just a change to a few lines to enable an additional `special` attribute besides "cancel".  I can provide the code for both the patch module and the better approach of adding a "no_save" option for the "special" attribute.  

I would urge you to consider doing this because of the utility of in-edit buttons and because the current behavior is fundamentally flawed in not respecting the user's choice of when to commit edits.

**Comment by elgow** [comment]
@ged-odoo I've fixed an issue with my CLA signature. Once it's done I can submit a pull request for the changes needed to enable a "no_save" option for the "special" attribute of buttons that enables in-edit button actions.

**Comment by elgow** [comment]
@ged-odoo @Yenthe666  I understand what ged-odoo said above about ensuring that server-side actions have up-to-date data in the DB to act upon. In the same form that has my in-edit buttons I have several buttons that perform server-side actions. What I've done is to make my in-edit buttons `oe_edit_only` and my server-side buttons `oe_read_only`  so  that the user cannot use the buttons when the DB state is wrong. This would be a lot of work if one tried to retrofit it to the Odoo code base, but if the button behavior remains as it is now with only an option for in-edit use then the work is really quite minimal.

**Comment by elgow** [comment]
I've submitted a new issue #21012 that is related to this one. The code that I wrote to solve the breaking change described in the new issue also prevents the save described in this one.  For in-edit button actions it's best if both the save and the reload are prevented.
